### ISSUE TITLE
:meth: does not properly consider current class in a multiple module scenario

### ISSUE THREAD

**Issue by Rapptz** [sphinx-doc/sphinx#3739]
Subject: Some cross references are hard to do or impossible when using autodoc with inherited-members and `:meth:`.

### Problem

This problem is a bit hard to describe since it involves using two different modules.

I am not sure if it intended behaviour, but I found the behaviour to be a little bit surprising and unintuitive with workarounds being a bit difficult to do.

Basically when using multiple modules, e.g. `test` and `test.extra` and having something in `test.extra` inherit from something in `test`, cross referencing using `:meth:` breaks when using `.. currentmodule:: test` or no current module set.

#### Procedure to reproduce the problem

Consider a structure like the following:

![](https://this.is-serious.business/81ab22.png)

with the contents of `a.py` being:

```py
class A:
    """This is a test"""

    def foo(self):
        """Does something."""
        pass

    def baz(self):
        """Does something similar to :meth:`foo`."""
        pass
```

and the contents of `b.py` being:

```py
import test.a

class B(test.a.A):
    """This is also a test."""
    def test(self):
        """Does something completely unrelated to :meth:`foo`"""
        pass
```

Having a basic `index.rst` with the following contents:

```rst
.. currentmodule:: test

.. autoclass:: test.A
    :members:

.. autoclass:: test.extra.B
    :members:
    :inherited-members:
```

#### Error logs / results

Causes the rendering to break cross references in `test.extra.B`: 

![](https://this.is-serious.business/20cbb4.png)

Changing the `:meth:obj` to use the `.` prefix breaks it as well leading to the obvious ambiguity warning:

```
D:\sphinx-bug-test\docs\index.rst:11: WARNING: more than one target found for cross-reference 'foo': test.A.foo, test.extra.B.foo
D:\sphinx-bug-test\docs\index.rst:23: WARNING: more than one target found for cross-reference 'foo': test.A.foo, test.extra.B.foo
```

#### Expected results

A way to make it link to the current class somehow. Ergo, the `foo` links in `test.extra.B` would link to `test.extra.B.foo` and in `test.A` links it would link to `test.A.foo`.

If this is some intended behaviour then forgive me. I'm interested in finding out how to solve my problem though, if it's even possible.

### Reproducible project / your project

[sphinx-bug-test.zip](https://github.com/sphinx-doc/sphinx/files/994997/sphinx-bug-test.zip)

### Environment info
- OS: Windows 8.1
- Python version: 3.5.2
- Sphinx version: 1.5.5

**Comment by tk0miya** [comment]
Thank you for reporting.

In my short investigation:
`.. autoclass:: test.extra.B` is expanded to following notation internally:
```
.. py:class:: B
   :module: test.extra

   This is also a test.


   .. py:method:: B.baz()
      :module: test.extra

      Does something similar to :meth:`foo`.


   .. py:method:: B.foo()
      :module: test.extra

      Does something.


   .. py:method:: B.test()
      :module: test.extra

      Does something completely unrelated to :meth:`foo`
```

As a result, `:meth:'foo'` is recognized as following:
* module: test (not `test.extra`, this comes from `currentmodule` directive)
* class: B
* reference target: foo

I don't know why autodoc ignores module name of `B` class. I feel it's bug of the autodoc (or python domain?).
I will take a look more deeper later.

Note: As a workaround, you can build your docs correctly with following notation:
```
.. currentmodule:: test

.. autoclass:: test.A
    :members:

.. currentmodule:: test.extra

.. autoclass:: test.extra.B
    :members:
    :inherited-members:
```

**Comment by tk0miya** [comment]
Fixed by #4176 
Thank you for reporting
